core/domain/user_services.py

```diff
@@ -697,7 +815,7 @@ def has_ever_registered(user_id):
     return bool(user_settings.username and user_settings.last_agreed_to_terms)
 
 
-def has_fully_registered_account(user_id):
+def has_fully_registered_account(user_id: Optional[str]) -> bool:
```

Comment:
Ditto.

Reply:
Here, `strict=true` is used so if we provide user_id as `None` with check for None case then it will raise error rather than returning `None`

Reply:
So, here i updated the docstring !!

Reply:
I think this is odd. It's probably better to prevent callsites from ever passing None to this function, right?

Reply:
Optional is removed from here with some refactoring in the function body. But I have encountered a test case where `None` is used for guest user. 
https://github.com/oppia/oppia/blob/1bd5ab86be391bb9ee896cfe0514ab3ab8093665/core/domain/user_domain_test.py#L376
So, I'm not sure removing `None` from here is a good thing or not!!!

Reply:
Thanks, good to check this. Is has_fully_registered() ever called with None outside a test context?

If not, we can drop the test.

If so, then we should fix those places to not call it with None (and then drop the test).

Reply:
(Or I suppose, in both cases, you could still keep the test for "defense in depth".)

Reply:
https://github.com/oppia/oppia/blob/c5c4515b361f0b21494ab4f7333ef9610665060c/core/controllers/profile.py#L333
https://github.com/oppia/oppia/blob/c5c4515b361f0b21494ab4f7333ef9610665060c/core/controllers/profile.py#L425

Here I added assertion!! 
Because the type of `self.user_id` is Optional[str] and I didn't find any test case where user_id is provided as None.

Reply:
https://github.com/oppia/oppia/blob/c5c4515b361f0b21494ab4f7333ef9610665060c/core/controllers/profile.py#L333
https://github.com/oppia/oppia/blob/c5c4515b361f0b21494ab4f7333ef9610665060c/core/controllers/profile.py#L425

Here In both places, I added assertion!!
Because the type of `self.user_id` is Optional[str] and I didn't find any case where `User_id` is provided as None.

Reply:
I also tried to hit the `SignupPage` handler as a guest user ( by going to `//signup`). But i'm only able to do it when I logged in and do not able to do it as a guest user. So I assumed that user_id I not going to be none. But for extra security i added assertion here ( if you prefer also ).